### ISSUE TITLE
Connections issue Pushing to corporate Docker Registry using JIB Plugin

### ISSUE THREAD

**Issue by khande-incomm** [GoogleContainerTools/jib#1966]
**Environment**: 

- *Jib version:* 1.5.1
- *Build tool:* Maven 3.2.1
- *OS:* Windows 10
-*Docker Version:* Docker Desktop v19.03.1

**Description of the issue**: 
I'm trying to deploy Docker image of my Spring Boot application to a private registry using Google JIB maven plugin. However, it fails with issues accessing the private registry.
I have added the CERT to 'example.registry.com' (could not give the actual registry as the one I'm using is corporate registry which is not available over the public internet) to the Windows trusted CERTS. I was able to verify this using the docker command to login to private registry as below :-

**docker login example.registry.com**

This command returns success for authentication which means the CERT was accepted.

And Following the logon to the private registry, I'm able to PULL and PUSH images using docker command line to 'example.registry.com' successfully .

However, using JIB plugin in my project's POM.xml and then execution command 'mvn compile jib:build' causes failures.

Please suggest if I'm making any mistake or missing any configuration.

**Expected behavior**:
mvn compile jib:build should complete successfully.

**Steps to reproduce**:

  1. Built Spring Boot Project
  2. Added JIB Pluging with POM file as followed
  3. run 'mvn compile jib:build'
  4. Build Failure

**`jib-maven-plugin` Configuration**: 
```xml
<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>jib-maven-plugin</artifactId>
				<version>1.5.1</version>
				<configuration>
					<to>
                        <image>example.registry.com/inventory-service-docker:1.0</image>
                        <auth>
					      <username>myuser</username>
					      <password>userpass</password>
			</auth>
                    </to>
                    <from>                           
				        <image>example.registry.com/openjdk:8</image>
				    </from>
                    <container>
        				<ports>
				            <port>8089</port>
				        </ports>
				    </container>
				   <!--  <allowInsecureRegistries>true</allowInsecureRegistries> -->
				</configuration>
			</plugin>
```

**Log output**: 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 7.759 s
[INFO] Finished at: 2019-09-09T12:28:46-05:00
[INFO] Final Memory: 32M/292M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:build (default-cli) on project inventory-service-docker: Build image failed, perhaps you should use a registry that supports HTTPS or set the configuration parameter 'allowInsecureRegistries': Failed to verify the server at https://example.registry.com/v2/inventory-service-docker/blobs/sha256:9cc2ad81d40d54dcae7fa5e8e17d9c34e8bba3b7c2cc7e26fb22734608bda32e because only secure connections are allowed.

**Additional Information**: Setting  'allowInsecureRegistries' to true as suggested on the first error also causes the build to fail but with a CERT error . Below is the POM file :-

**`jib-maven-plugin` Configuration**: 
```xml
<plugin>
       <groupId>com.google.cloud.tools</groupId>
	<artifactId>jib-maven-plugin</artifactId>
	<version>1.5.1</version>
	<configuration>
	<to>
                    <image>example.registry.com/inventory-service-docker:1.0</image>
        </to>
        <from>                           
	     <image>example.registry.com/openjdk:8</image>
	 </from>
         <container>
        	<ports>
	            <port>8089</port>
	        </ports>
	   </container>
	   <allowInsecureRegistries>true</allowInsecureRegistries>
	</configuration>
</plugin>
```
**Log output**:

[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.740 s
[INFO] Finished at: 2019-09-09T15:13:02-05:00
[INFO] Final Memory: 79M/211M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:build (default-cli) on project inventory-service-docker: Build image failed: Failed to authenticate with registry example.registry.com/inventory-service-docker because: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

I tried adding Trust Store to the exmaple.registry.com' also on maven command as below , but it still fails with same CERT error

mvn compile -Djavax.net.ssl.trustStore=C:\Java\jdk1.8.0_40\jre\lib\security\DockerTrustStore -Djavax.net.ssl.trustStorePassword=mydockerpass jib:build

**Comment by chanseokoh** [comment]
Hi @khande-incomm,

I'm no expert in Windows, but this [doc](https://www.benecke.cloud/java-runtime-environment-jre-how-to-import-certificates/) says,

> One big problem of the JRE is that it completely ignores the Windows certificate stores. Instead of using the windows certificate store it uses its own implementation.

The first failure without `<allowInsecureRegistries>` might be because of this. Have you tried `mvn -Djavax.net.ssl.trustStore=C:\Java\jdk1.8.0_40\jre\lib\security\DockerTrustStore ...` *without* `<allowInsecureRegistries>`?

And double-check `C:\Java\jdk1.8.0_40\jre\lib\security\DockerTrustStore` has your cert. Using [KeyStore Explorer](http://keystore-explorer.org/), for example.

Also test whether the JVM successfully picks up `-Djavax.net.ssl.trustStore=C:\...\DockerTrustStore` with a [simple Java program](https://gist.github.com/4ndrej/4547029#gistcomment-727201) (However, I have never tried `SSLPoke` myself.)

And I think the second failure when `<allowInsecureRegistries>true` is set is arguably "working as intended. This is when the connection to an auth server is not verifiable, and we intentionally do not allow insecure connection to an auth server even if `<allowInsecureRegistries>` is set. In your case, once Java can pick up your cert correctly, you should be fine without `<allowInsecureRegistries>`. (FTR, this is from `RegistryAuthenticator.authenticate()`.)

**Comment by chanseokoh** [comment]
@khande-incomm any update on this? Did you manage to make it work?

Not directly related to this, but I see you're using Jib 1.5.1, which we found to have an issue of very slow network operations. You should upgrade to 1.6.0.

**Comment by khande-incomm** [comment]
@chanseokoh ,

Thanks for the response and pointing me in right direction, me adding the CERT of my private registry to the JAVA CACERT file did the trick. Adding the CERT, JIB was successfully able to validate my credentials on the private registry and then pushed the image to it. I  used following sample to add the CERT.

`keytool.exe -import 
-storepass "changeit" 
-keystore "C:\Program Files (x86)\Java\jre1.8.0_144\lib\security\cacerts" 
-alias certificate.cer 
-file "C:\certificate.cer" 
-noprompt`

Although, to your other point from earlier post, I had tried using the `DockerTrustStore` as a parameter to `mvn compile jib:build` I had created  using the same CERT, but that did not work. 

So I guess it has to be the CERT.

Thanks
Kaustubh
